design/kubernetes-friendly-oam.md

```diff
@@ -0,0 +1,482 @@
+are easily discoverable, and allows the OAM runtime to inject parameters into
+any field of the embedded workload without introducing a [“fromParams()” inline
+function DSL]. This approach also allows the published parameters to omit their
+type - the type can be inferred from the type of the workload fields a parameter
+corresponds to.
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: WorkloadDefinition
+metadata:
+  name: rediscluster.cache.crossplane.io
+spec:
+  definitionRef:
+    name: rediscluster.cache.crossplane.io
+```
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: Component
+metadata:
+  name: example-cache
+spec:
+  workload:
+    apiVersion: cache.crossplane.io/v1alpha1
+    kind: RedisCluster
+    spec:
+      engineVersion: "3.2"
+  parameters:
+  - name: instanceName
+    required: true
+    fieldPaths:
+    - ".metadata.name"
+  - name: secret
+    required: true
+    fieldPaths:
+    - ".spec.writeConnectionSecretToRef.name"
+  - name: engineVersion
+    required: false
+    fieldPaths:
+    - ".spec.engineVersion"
+```
+
+The above WorkloadDefinition and Component leverages the existing Crossplane
+RedisCluster resource claim. Crossplane as a tentative OAM runtime need not
+define an extended workload type; it need only create a WorkloadDefinition
+pointing to [the existing CustomResourceDefinition] of the RedisCluster resource
+claim.
+
+The Component design enables OAM runtimes to break the task of reconciling the
+desired state of an application with the actual state of the world into two
+distinct stages:
+
+1. Render workloads. The OAM runtime iterates over the Components referenced by
+   an ApplicationConfiguration, instantiating a distinct workload resource for
+   each component. In the above examples this would involve creating a distinct
+   Server resource, and a distinct RedisCluster resource.
+2. Reconcile workloads. By producing valid instances of existing Kubernetes
+   custom resources OAM is able to leverage existing controllers. For example
+   the existing Crossplane RedisClaim controller(s) would be responsible for
+   reconciling the RedisCluster custom resource produced during the render
+   phase.
+
+Note that this two phase reconciliation and the production of distinct workload
+resources would not be required by the OAM specification. Runtimes could choose
+to iterate over the referenced components, render them in memory, and act upon
+the rendered components directly.
+
+### Traits
+
+Like workloads, each trait is of a distinct kind. A TraitDefinition defines the
+schema of each kind of trait, and limits what workload kinds it applies to:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: TraitDefinition
+metadata:
+  name: manualscalertrait.core.oam.dev
+spec:
+  appliesToWorkloads:
+    - server.core.oam.dev/v1alpha1
+    - worker.core.oam.dev/v1alpha1
+    - task.core.oam.dev/v1alpha1
+  definitionRef:
+    name: manualscalertrait.core.oam.dev
+```
```

Comment:
besides definitionRef, we'd better add `definition` field here. Then we could allow non-K8s implementation to define json schema here.

Reply:
``` yaml
apiVersion: core.oam.dev/v1alpha1
kind: TraitDefinition
metadata:
  name: manualscalertrait.core.oam.dev
spec:
  appliesToWorkloads:
    - server.core.oam.dev/v1alpha1
    - worker.core.oam.dev/v1alpha1
    - task.core.oam.dev/v1alpha1
  definitionRef:
    name: manualscalertrait.core.oam.dev
  definition: |
      {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "type": "object",
      "required": [
        "replicaCount"
      ],
      "properties": {
        "replicaCount": {
          "type": "integer",
          "description": "the target number of replicas to scale a component to.",
          "minimum": 0
        }
      }
    }
```

Reply:
after discuss with @ryanzhang-oss , we both agree that non-k8s implementation can also use definitionRef. So we'd better not add `definition`.